File: apps/api/liderix_api/schemas/teams.py
from __future__ import annotations

from datetime import datetime
from typing import Optional, List
from uuid import UUID

from pydantic import BaseModel, ConfigDict

from liderix_api.enums import MembershipRole


class TeamMemberRead(BaseModel):
    id: UUID
    name: str
    email: str
    role: MembershipRole
    department: Optional[str] = None
    position: Optional[str] = None
    avatar_url: Optional[str] = None
    join_date: Optional[datetime] = None
    status: Optional[str] = None
    tasks_completed: int = 0
    projects_active: int = 0
    tasks_open: int = 0
    tasks_overdue: int = 0
    tasks_in_review: int = 0
    tasks_blocked: int = 0

    model_config = ConfigDict(extra="forbid")


class TeamRead(BaseModel):
    id: UUID
    name: str
    description: Optional[str] = None
    department: Optional[str] = None
    lead: Optional[str] = None
    policy: Optional["TeamPolicy"] = None
    members: List[TeamMemberRead] = []
    projects: int = 0
    tasks_open: int = 0
    tasks_overdue: int = 0
    tasks_in_review: int = 0
    tasks_blocked: int = 0
    created_at: Optional[datetime] = None

    model_config = ConfigDict(extra="forbid")


class TeamListResponse(BaseModel):
    items: List[TeamRead]
    total: int

    model_config = ConfigDict(extra="forbid")


class TeamPolicy(BaseModel):
    default_approver_role: Optional[str] = None
    escalation_days: Optional[int] = None
    weekly_digest_recipients: Optional[List[str]] = None

    model_config = ConfigDict(extra="forbid")

File: n8n/src/lib/api/teams.ts
import { api } from "@/lib/api/config"
import { CompanyAPI } from "@/lib/api/company"
import { MembershipRole } from "@/types/roles"

export interface TeamMember {
  id: string
  name: string
  email: string
  role: MembershipRole
  department?: string | null
  position?: string | null
  avatar_url?: string | null
  join_date?: string | null
  status?: string | null
  tasks_completed: number
  projects_active: number
  tasks_open: number
  tasks_overdue: number
  tasks_in_review: number
  tasks_blocked: number
}

export interface TeamPolicy {
  default_approver_role?: MembershipRole
  escalation_days?: number
  weekly_digest_recipients?: string[]
}

export interface Team {
  id: string
  name: string
  description?: string | null
  department?: string | null
  lead?: string | null
  policy?: TeamPolicy | null
  members: TeamMember[]
  projects: number
  tasks_open: number
  tasks_overdue: number
  tasks_in_review: number
  tasks_blocked: number
  created_at?: string | null
}

export interface TeamListResponse {
  items: Team[]
  total: number
}

async function getDefaultOrgId(): Promise<string | null> {
  const org = await CompanyAPI.getCurrentCompany()
  return org?.id ?? null
}

export async function getTeams(orgId?: string): Promise<Team[]> {
  const resolvedOrgId = orgId ?? (await getDefaultOrgId())
  if (!resolvedOrgId) return []
  const { data } = await api.get<TeamListResponse>(`/orgs/${resolvedOrgId}/teams/`)
  return data.items ?? []
}

export async function createTeam(payload: {
  name: string
  description?: string
  manager_id?: string
  policy?: TeamPolicy
}): Promise<void> {
  const orgId = await getDefaultOrgId()
  if (!orgId) return
  await api.post(`/orgs/${orgId}/departments/`, {
    name: payload.name,
    description: payload.description,
    manager_id: payload.manager_id,
    policy: payload.policy,
  })
}

export async function inviteTeamMember(payload: {
  email: string
  role: MembershipRole
  department_id?: string
}): Promise<void> {
  const orgId = await getDefaultOrgId()
  if (!orgId) return
  await api.post(`/orgs/${orgId}/invitations`, {
    invited_email: payload.email,
    role: payload.role,
    department_id: payload.department_id,
  })
}
